Lib/typing.py

```diff
@@ -1393,7 +1393,11 @@ def __mro_entries__(self, bases):
                 return ()
             i = bases.index(self)
             for b in bases[i+1:]:
-                if isinstance(b, _BaseGenericAlias) and b is not self:
+                if (
+                    (isinstance(b, _BaseGenericAlias)
+                        or Generic in getattr(b, '__mro__', ()))
+                    and b is not self
+                ):
```

Comment:
Hmm, looks like I broke a test with this suggestion, but I can't really say I understand how.

Reply:
This happens because in a case of `class L2(Generic[S], L[S]): ...` we would get `b` as `L[S]` which is a `GenericAlias`, it is not a `type`. So, your condition would be `False`. But, it still has `__mro__`, so the original one will work correctly. 

We can also add 

```python
if isinstance(b, GenericAlias):
     return ()
```

But, it does not feel safe to me.

Reply:
Could you elaborate on that? As far as I can see `L2` is currently defined as inheriting from `List`, which is a `typing._GenericAlias`. Conversely, the built-in `list` is a `types.GenericAlias`. And both are actually a `type` as far as I can tell.

I played around with it a bit on Python 3.11.3 (not this branch):
```ipython
>>> import typing
>>> T = typing.TypeVar("T")
>>> class LowercaseList(list[T]): pass
...
>>> class UppercaseList(typing.List[T]): pass
...
>>> LowercaseList.__mro__
(<class '__main__.LowercaseList'>, <class 'list'>, <class 'object'>)
>>> UppercaseList.__mro__
(<class '__main__.UppercaseList'>, <class 'list'>, <class 'typing.Generic'>, <class 'object'>)
>>> issubclass(LowercaseList, typing.Generic)
False
>>> issubclass(UppercaseList, typing.Generic)
True
>>> isinstance(LowercaseList, type)
True
>>> isinstance(UppercaseList, type)
True
>>> import types
>>> isinstance(types.GenericAlias, type)
True
```

That said, I did notice that the built-in `list` does not have `Generic` in its bases (see `LowercaseList.__mro__` above). So I don't see why we would want to `return ()` for them in the first place as that would just result in the absence of `Generic` in the MRO. What happens when you just use my initial suggestion? Does it actually raise a `TypeError`? For which types?

Reply:
Indeed, I see. `types.GenericAlias` is not a subclass of `type` but it is supported in `issubclass` so that adding the `isinstance(b, type)` check changes the behavior.

Do I understand correctly that the patch as I initially proposed it doesn't actually result in a `TypeError` (except where the stable implementation does as well) because it doesn't really change when `issubclass` is called, and that your change is more a precaution against something that affects the current stable Python as well, i.e. more or less disjunct from #106102?

How about just trying the `issubclass` and catching the `TypeError`? It is generic in the sense that it doesn't assume anything about which objects are allowed in an `issubclass` check, unlike the `isinstance(b, type)` guard.

```python
try:
    if issubclass(b, Generic):
       return ()
except TypeError:
    continue
```

Reply:
@sanderr I tried it, didn't work :)

```
======================================================================
ERROR: test_multiple_inheritance_special_multiple_layers (test.test_typing.GenericTests.test_multiple_inheritance_special_multiple_layers)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython/Lib/test/test_typing.py", line 4878, in test_multiple_inheritance_special_multiple_layers
    class L2(Generic[S], L[S]): ...
TypeError: Cannot create a consistent method resolution order (MRO) for bases Generic, L

----------------------------------------------------------------------
Ran 595 tests in 0.193s

FAILED (errors=1, skipped=1)
test test_typing failed
test_typing failed (1 error)

== Tests result: FAILURE ==

1 test failed:
    test_typing

Total duration: 321 ms
Tests result: FAILURE
```